Track each SST's timestamp information as user properties

sunlike-Lipo opened the pull request:
Track each SST's timestamp information as user properties #8959

Rockdb has supported user-defined timestamp feature. Application can specify a timestamp
when writing each k-v pair. When data flush from memory to disk file called SST files. 
Each SST files consist of multiple data blocks and several metadata blocks. Among the metadata 
blocks, there is one called Properties block that tracks some pre-defined properties of this SST file.

This PR is for collecting the properties of min and max timestamps of all keys in the file. With those
properties the SST file is more convenient to tell whether the keys in the SST have timestamps or not.

The changes involved are as follows:

1) Add a class TimestampTablePropertiesCollector to collect min/max timestamp when add keys to table,
   The way TimestampTablePropertiesCollector use to compare timestamp of key should defined by
   user by implementing the Comparator::CompareTimestamp function in the user defined comparator.
2) Add corresponding unit tests.

wolfkdy commented:
@riversand963  I've reviewed this mr and my colleague post it.  
The original mr is splitted into two mrs.
What you commented in the original mr are all fixed.
Shall you take a look when free, thanks!

riversand963 commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,60 @@ class UserKeyTablePropertiesCollectorFactory
+
+  Status InternalAdd(const Slice& key, const Slice& /* value */,
+                     uint64_t /* file_size */) override {
+    if (!cmp_) {
+      return Status::OK();
+    }
+    auto timestampInKey =
+        ExtractTimestampFromUserKey(key, cmp_->timestamp_size());
+    if (timestampInKey == kDisableUserTimestamp) {
+      return Status::OK();
+    }
```
Didn't check thoroughly, but can `cmp_` be null here?

Can this be
```
assert (cmp_);
if (cmp_->timestamp_size() == 0) {
  return Status::OK();
}
if (key.size() < cmp_->timestamp_size()) {
  return some error;
}
```

sunlike-Lipo replied:
fine. Already fixed it.

sunlike-Lipo replied:
Since now adding TimestampTablePropertiesCollector only when timestamp enabled, here assert change to

assert (cmp_ && cmp_->timestamp_size() > 0);

riversand963 commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,60 @@ class UserKeyTablePropertiesCollectorFactory
+  }
+
+  Status Finish(UserCollectedProperties* properties) override {
+    properties->insert({"min_timestamp", min_timestamp});
```
Can we check the size of min/max timestamps are both equal to cmp_->timestamp_size()?

sunlike-Lipo replied:
fine. Already fixed it.

riversand963 commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,60 @@ class UserKeyTablePropertiesCollectorFactory
+  }
+
+  UserCollectedProperties GetReadableProperties() const override {
+    // left blank
```
What does this comment mean? No real implementation for this override method?

pdillinger replied:
overrides pure virtual

sunlike-Lipo replied:
Yeah, this is for overrides pure virtual fuction.  Already removed this comment.

riversand963 commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,60 @@ class UserKeyTablePropertiesCollectorFactory
+
+ protected:
+  const Comparator* cmp_;
+  std::string max_timestamp;
+  std::string min_timestamp;
```
`max_timestamp_` and `min_timestamp_` to follow the same naming convention of member variables.

sunlike-Lipo replied:
Sure. Already fixed it

riversand963 commented on table/block_based/block_based_table_builder.cc:
```diff
@@ -515,6 +515,9 @@ struct BlockBasedTableBuilder::Rep {
         new BlockBasedTablePropertiesCollector(
             table_options.index_type, table_options.whole_key_filtering,
             moptions.prefix_extractor != nullptr));
+    table_properties_collectors.emplace_back(
```
Can we do this only if timestamp is enabled? Then in the TimestampTablePropertiesCollector, we can always assert timestamp_size > 0

pdillinger replied:
+1  I would expect a non-trivial performance hit as it is.

sunlike-Lipo replied:
Sure. Already fixed it

riversand963 commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,60 @@ class UserKeyTablePropertiesCollectorFactory
+  }
+
+ protected:
+  const Comparator* cmp_;
```
Do you want this class to be assignable? if not, we can do
```
const Comparator* const cmp_;
```

sunlike-Lipo replied:
Sure. Already fixed it

riversand963 reviewed (commented):
Thanks @sunlike-Lipo for the PR. I think this PR needs more testing once #9092 gets merged. cc @wolfkdy

wolfkdy commented:
@riversand963 
> Thanks @sunlike-Lipo for the review. I think this PR needs more testing once #9092 gets merged. cc @wolfkdy
Thanks, we plan to follow other unittests for TablePropertiesCollector's subclass and add unittests for TimestampTablePropertiesCollector.

sunlike-Lipo commented:
@riversand963 
I fixed review comments and add a unitTest.  Could you please help me to review what else needs to be modified if that is convenient.

riversand963 commented:
@sunlike-Lipo Will take a look.
One question: do you by any chance know the overhead introduced by tracking min/max timestamps for each file, since the comparison will be done for each key? It would be nice to include some benchmarking results here. We can assume timestamp size of 8 bytes.

riversand963 reviewed (commented):
A few more comments.

sunlike-Lipo commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,64 @@ class UserKeyTablePropertiesCollectorFactory
+
+  Status InternalAdd(const Slice& key, const Slice& /* value */,
+                     uint64_t /* file_size */) override {
+    auto uKey = ExtractUserKey(key);
```
OK, fixed.

sunlike-Lipo commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,64 @@ class UserKeyTablePropertiesCollectorFactory
+      return Status::NotSupported(
+          "Not supported to extract timestamp from user key.");
+    }
+    auto timestampInKey =
```
OK, fixed.

sunlike-Lipo commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,64 @@ class UserKeyTablePropertiesCollectorFactory
+    }
+    auto timestampInKey =
+        ExtractTimestampFromUserKey(uKey, cmp_->timestamp_size());
+    if (timestampInKey == kDisableUserTimestamp) {
```
I forgot to remove it.  removed the if section.
sorry for my carelessness.

sunlike-Lipo commented on db/dbformat.h:
```diff
@@ -91,6 +91,8 @@ static const SequenceNumber kMaxSequenceNumber = ((0x1ull << 56) - 1);
 
 static const SequenceNumber kDisableGlobalSequenceNumber = port::kMaxUint64;
 
+static const std::string kDisableUserTimestamp = "";
```
removed

sunlike-Lipo commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,64 @@ class UserKeyTablePropertiesCollectorFactory
+  void BlockAdd(uint64_t /* block_raw_bytes */,
+                uint64_t /* block_compressed_bytes_fast */,
+                uint64_t /* block_compressed_bytes_slow */) override {
+    // left blank
```
removed

sunlike-Lipo commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,64 @@ class UserKeyTablePropertiesCollectorFactory
+    auto uKey = ExtractUserKey(key);
+    assert(cmp_ && cmp_->timestamp_size() > 0);
+    if (uKey.size() < cmp_->timestamp_size()) {
+      return Status::NotSupported(
```
fix it to Corruption("User key size mismatch when comparing to timestamp size.")

sunlike-Lipo commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,64 @@ class UserKeyTablePropertiesCollectorFactory
+  }
+
+  UserCollectedProperties GetReadableProperties() const override {
+    return UserCollectedProperties();
```
added.

sunlike-Lipo commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,64 @@ class UserKeyTablePropertiesCollectorFactory
   std::shared_ptr<TablePropertiesCollectorFactory> user_collector_factory_;
 };
 
+class TimestampTablePropertiesCollector : public IntTblPropCollector {
```
added some decription

sunlike-Lipo commented:
> @sunlike-Lipo Will take a look. One question: do you by any chance know the overhead introduced by tracking min/max timestamps for each file, since the comparison will be done for each key? It would be nice to include some benchmarking results here. We can assume timestamp size of 8 bytes.

Sure. I'll do the test and post the benchmarking results as soon as possible.

sunlike-Lipo commented:
> @sunlike-Lipo Will take a look. One question: do you by any chance know the overhead introduced by tracking min/max timestamps for each file, since the comparison will be done for each key? It would be nice to include some benchmarking results here. We can assume timestamp size of 8 bytes.


I wrote a unittest(https://github.com/sunlike-Lipo/rocksdb/commit/d61cbab3dfbb3109378adca04dced8c0aaf24a3b) for the benchmark,
In general, the test runs for 10 times and collects the total time used for sstfiles' Flush.
I use the MockEnv to eliminate disk-IO's effect.

The variables are `value size` and `collect timestamp or not`,. The comparator used is ComparatorWithU64Ts.
the test results are list below

|                | value_size=10 | value_size=100 | value_size=1000 |
| -------------- | ------------- | -------------- | --------------- |
| collect-ts     | 4777ms        | 1517ms         | 644ms           |
| not-collect-ts | 4305ms        | 1458ms         | 619ms           |
| downgrade      | 10.9%         | 4.0%           | 4.0%            |

sunlike-Lipo commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,64 @@ class UserKeyTablePropertiesCollectorFactory
+  explicit TimestampTablePropertiesCollector(const Comparator* cmp)
+      : cmp_(cmp) {}
+
+  virtual ~TimestampTablePropertiesCollector() {}
```
fixed

riversand963 commented:
Thanks @sunlike-Lipo for addressing the comments. Will take another look soon.

> I wrote a unittest

I wonder if we can move the code to a standalone source file (similar to examples/) and compile with rocksdb library built with `DEBUG_LEVEL=0` and see the performance difference in opt mode too. You can also use `db_bench`. In unit test, I think the DEBUG_LEVEl is at least 1, thus may already have some overhead, making the performance overhead of ts comparison less obvious.

sunlike-Lipo commented:
> Thanks @sunlike-Lipo for addressing the comments. Will take another look soon.
> 
> > I wrote a unittest
> 
> I wonder if we can move the code to a standalone source file (similar to examples/) and compile with rocksdb library built with `DEBUG_LEVEL=0` and see the performance difference in opt mode too. You can also use `db_bench`. In unit test, I think the DEBUG_LEVEl is at least 1, thus may already have some overhead, making the performance overhead of ts comparison less obvious.

I fixed it to a stand alone test (https://github.com/sunlike-Lipo/rocksdb/commit/50a1152e1172b239036a096ddadc2d24ac338521) for the benchmark. The logical is not changed:
In general, the test runs for 10 times and collects the total time used for sstfiles' Flush.
I use the MemEnv to eliminate disk-IO's effect.

The variables are `value size` and `collect timestamp or not`,. The comparator still is ComparatorWithU64Ts.
the test results are list below

|                | value_size=10 | value_size=100 | value_size=1000 |
| -------------- | ------------- | -------------- | --------------- |
| collect-ts     | 4694ms        | 1486ms         | 652ms           |
| not-collect-ts | 4319ms        | 1464ms         | 634ms           |
| downgrade      | 8.6%          | 1.5%           | 2.8%            |

I will also test the db_bench, the result will be soon.

sunlike-Lipo commented:
> Thanks @sunlike-Lipo for addressing the comments. Will take another look soon.
> 
> > I wrote a unittest
> 
> I wonder if we can move the code to a standalone source file (similar to examples/) and compile with rocksdb library built with `DEBUG_LEVEL=0` and see the performance difference in opt mode too. You can also use `db_bench`. In unit test, I think the DEBUG_LEVEl is at least 1, thus may already have some overhead, making the performance overhead of ts comparison less obvious.

db_bench test is here:
command:
```
$ TEST_TMPDIR=/tmp/rocksdb ./db_bench -benchmarks=filluniquerandom -level0_slowdown_writes_trigger 102400 -level0_stop_writes_trigger 102400  -disable_auto_compactions true -user_timestamp_size 8 -compression_type=none  -num=20000000
```
With ts collector       : 22.8MB/s, 22.5MB/s, 22.5MB/s, 22.6MB/s, 22.6MB/s
With out ts collector : 22.5MB/s, 22.3MB/s, 22.6MB/s, 22.7MB/s, 22.6MB/s

The detailed results are as follows:
With ts collector     : 
5.193 micros/op 192573 ops/sec 22.8  MB/S
5.263 micros/op 190004 ops/sec 22.5  MB/S  
5.266 micros/op 189885 ops/sec 22.5  MB/S
5.242 micros/op 190750 ops/sec 22.6  MB/S
5.228 micros/op 191274 ops/sec 22.6  MB/S
                                                                                                                                             
With out ts collector : 
5.256 micros/op 190257 ops/sec 22.5  MB/s
5.298 micros/op 188735 ops/sec 22.3 MB/s
5.229 micros/op 191234 ops/sec 22.6 MB/s
5.204 micros/op 192158 ops/sec 22.7 MB/s
5.233 micros/op 191098 ops/sec 22.6 MB/s

riversand963 commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,65 @@ class UserKeyTablePropertiesCollectorFactory
+// internal key when Add() is invoked.
+//
+// @param cmp the Comparator to compare timestamp of key by the
+//     Comparator::CompareTimestamp function.
```
Not sure, but the indentation here seems weird.

sunlike-Lipo replied:
fixed

riversand963 commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,65 @@ class UserKeyTablePropertiesCollectorFactory
+  Status Finish(UserCollectedProperties* properties) override {
+    assert(min_timestamp_.size() == max_timestamp_.size() &&
+           max_timestamp_.size() == cmp_->timestamp_size());
+    properties->insert({"rocksdb.min_timestamp", min_timestamp_});
+    properties->insert({"rocksdb.max_timestamp", max_timestamp_});
```
I used `sst_dump` to dump the properties of generated SST files, and I can see
```
...
 # rocksdb.max_timestamp: 0x03000000000000000000000000000000
  # rocksdb.merge.operands: 0x00
  # rocksdb.min_timestamp: 0x03000000000000000000000000000000
```
This makes me wonder whether we can instead use rocksdb.timestamp_min and rocksdb.timestamp_high so that they can be next to each other.

sunlike-Lipo replied:
fixed it.

riversand963 commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,65 @@ class UserKeyTablePropertiesCollectorFactory
+  }
+
+  UserCollectedProperties GetReadableProperties() const override {
+    return {{"rocksdb.min_timestamp", Slice(min_timestamp_).ToString(true)},
+            {"rocksdb.max_timestamp", Slice(max_timestamp_).ToString(true)}};
```
Same as previous one.
Can we instead use rocksdb.timestamp_min and rocksdb.timestamp_high so that they can be next to each other?

sunlike-Lipo replied:
fixed it.

riversand963 commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,65 @@ class UserKeyTablePropertiesCollectorFactory
+
+ protected:
+  const Comparator* const cmp_;
+  std::string max_timestamp_;
+  std::string min_timestamp_;
```
Should these two be explicitly initialized to `kDisableTimestamp`?

sunlike-Lipo replied:
fixed it.

riversand963 commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,65 @@ class UserKeyTablePropertiesCollectorFactory
+        ExtractTimestampFromUserKey(user_key, cmp_->timestamp_size());
+    if (max_timestamp_ == kDisableUserTimestamp ||
+        cmp_->CompareTimestamp(timestamp_in_key, max_timestamp_) > 0) {
+      max_timestamp_ = timestamp_in_key.ToString();
```
If we do
```
max_timestamp_.assign(timestamp_in_key.data(), timestamp_in_key.size());
```
will it save us some string creation and copy caused by calling `ToString()`?

sunlike-Lipo replied:
fixed it.

riversand963 commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,65 @@ class UserKeyTablePropertiesCollectorFactory
+    }
+    if (min_timestamp_ == kDisableUserTimestamp ||
+        cmp_->CompareTimestamp(min_timestamp_, timestamp_in_key) > 0) {
+      min_timestamp_ = timestamp_in_key.ToString();
```
Ditto

sunlike-Lipo replied:
fixed it.

riversand963 commented on db/table_properties_collector.h:
```diff
@@ -108,4 +110,65 @@ class UserKeyTablePropertiesCollectorFactory
+class TimestampTablePropertiesCollector : public IntTblPropCollector {
+ public:
+  explicit TimestampTablePropertiesCollector(const Comparator* cmp)
+      : cmp_(cmp) {}
```
Maybe here we can explicitly initialize `max_timestamp_` and `min_timestamp_`.

sunlike-Lipo replied:
fixed it.

riversand963 reviewed (commented):
Thanks @sunlike-Lipo for the PR. Left some final minor comments before stamping.

sunlike-Lipo commented:
I made some tests again after changing string =  to string.assign

The db_bench result are as follows:
With ts collector :
5.175 micros/op 193250 ops/sec 22.9 MB/S
5.149 micros/op 194219 ops/sec 23.0 MB/S
5.223 micros/op 191443 ops/sec 22.6 MB/S
5.167 micros/op 193534 ops/sec 22.9 MB/S
5.150 micros/op 194183 ops/sec 23.0 MB/S

Without ts collector :
5.209 micros/op 191963 ops/sec 22.7 MB/s
5.201 micros/op 192256 ops/sec 22.7 MB/s
5.147 micros/op 194269 ops/sec 23.0 MB/s
5.230 micros/op 191211 ops/sec 22.6 MB/s
5.269 micros/op 189781 ops/sec 22.4 MB/s

The standalone test from https://github.com/sunlike-Lipo/rocksdb/commit/50a1152e1172b239036a096ddadc2d24ac338521 result are as follows:
use string.assign

|                | value_size=10 | value_size=100 | value_size=1000 |
| -------------- | ------------- | -------------- | --------------- |
| collect-ts     | 3851ms        | 1219ms         | 202ms           |
| not-collect-ts | 3745ms        | 1181ms         | 189ms           |

use string.operator=

|                | value_size=10 | value_size=100 | value_size=1000 |
| -------------- | ------------- | -------------- | --------------- |
| collect-ts     | 3868ms        | 1239ms         | 204ms           |
| not-collect-ts | 3767ms        | 1189ms         | 195ms           |

It seems there is little difference between the result using string.assign and using string.operator=, perhaps the compiler is 
clever enough to optmize this. 
Any way, I'll use string.assign because it ought to be faster than string.operator=.
